x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/FrozenSearchableSnapshotsIntegTests.java

```diff
@@ -448,4 +458,119 @@ public void testCreateAndRestorePartialSearchableSnapshot() throws Exception {
+        final MountSearchableSnapshotRequest req = new MountSearchableSnapshotRequest(
+            "index",
+            "repo",
+            "snap",
+            "index",
+            indexSettingsBuilder.build(),
+            Strings.EMPTY_ARRAY,
+            true,
+            MountSearchableSnapshotRequest.Storage.SHARED_CACHE
+        );
+
+        final RestoreSnapshotResponse restoreSnapshotResponse = client().execute(MountSearchableSnapshotAction.INSTANCE, req).get();
+        assertThat(restoreSnapshotResponse.getRestoreInfo().failedShards(), equalTo(0));
+        ensureSearchable("index");
+
+        // use a fixed client for the searches, as clients randomize timeouts, which leads to different cache entries
+        Client client = client();
+
+        final SearchResponse r1 = client.prepareSearch("index")
+            .setSize(0)
+            .setSearchType(SearchType.QUERY_THEN_FETCH)
+            .addAggregation(
+                dateHistogram("histo").field("f").timeZone(ZoneId.of("+01:00")).minDocCount(0).calendarInterval(DateHistogramInterval.MONTH)
+            )
+            .get();
+        assertSearchResponse(r1);
+
+        assertRequestCacheState(client(), "index", 0, 1);
+
+        // The cached is actually used
+        assertThat(
+            client().admin()
+                .indices()
+                .prepareStats("index")
+                .setRequestCache(true)
+                .get()
+                .getTotal()
+                .getRequestCache()
+                .getMemorySizeInBytes(),
+            greaterThan(0L)
+        );
+
+        for (int i = 0; i < 10; ++i) {
+            final SearchResponse r2 = client.prepareSearch("index")
+                .setSize(0)
+                .setSearchType(SearchType.QUERY_THEN_FETCH)
+                .addAggregation(
+                    dateHistogram("histo").field("f")
+                        .timeZone(ZoneId.of("+01:00"))
+                        .minDocCount(0)
+                        .calendarInterval(DateHistogramInterval.MONTH)
+                )
+                .get();
+            assertSearchResponse(r2);
+            assertRequestCacheState(client(), "index", i + 1, 1);
+            Histogram h1 = r1.getAggregations().get("histo");
+            Histogram h2 = r2.getAggregations().get("histo");
+            final List<? extends Histogram.Bucket> buckets1 = h1.getBuckets();
+            final List<? extends Histogram.Bucket> buckets2 = h2.getBuckets();
+            assertEquals(buckets1.size(), buckets2.size());
+            for (int j = 0; j < buckets1.size(); ++j) {
+                final Histogram.Bucket b1 = buckets1.get(j);
+                final Histogram.Bucket b2 = buckets2.get(j);
+                assertEquals(b1.getKey(), b2.getKey());
+                assertEquals(b1.getDocCount(), b2.getDocCount());
+            }
+        }
+    }
```

Comment:
Can you also test that the cache entries are removed when closing the shard (via the number of evictions or the memory size)?

Reply:
I wanted to test that, but couldn't come up with a good way to do it without reaching directly into the implementation of IndicesRequestCache. The reason it's hard is that the stats are tracked on a per-shard level (within IndexShard), and so vanish when the shard is shut down. I've added some basic infra to reach into the IndicesRequestCache directly and extract the relevant information in ea607eb2632.